errors/errors.go

```
@@ -0,0 +1,75 @@
+/*Package errors defines a package used for error handling in 0-Disk.
+It provides an error type that has the original error as the cause,
+more context to the error can be provided by wrapping the error with an additional message.
+The underlying error (cause) will then be preserved and can be fetched (and then checked)
+with the Cause function.
+*/
+package errors
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// New returns an error with provided message
+func New(msg string) error {
+	return errors.New(msg)
+}
+
+// Newf formats an error according to a format specifier
+func Newf(format string, args ...interface{}) error {
+	return errors.Errorf(format, args...)
+}
+
+// Wrap returns an error that is annotated with provided message
+// If error is nil, Wrap returns nil
+func Wrap(err error, msg string) error {
+	return errors.WithMessage(err, msg)
+}
+
+// Wrapf returns an error that is annotated with provided message
+// If err is nil, Wrapf returns nil.
+func Wrapf(err error, format string, args ...interface{}) error {
+	message := fmt.Sprintf(format, args...)
+	return errors.WithMessage(err, message)
+}
+
+// Cause returns the underlying cause of the error if possible.
+// If the error does not implement `Cause() error` it returns the full error
+func Cause(err error) error {
+	return errors.Cause(err)
+}
+
+// NewErrSlice returns an empty ErrorSlice
+func NewErrSlice() *ErrorSlice {
+	return new(ErrorSlice)
+}
+
+// ErrorSlice represents a slice of errors
+type ErrorSlice struct {
```

This type was very much like the `ErrorSlice` type I implemented a couple of times in 0-Disk. However it always left me with a questioning feeling if this type is really good as it is.

One of the short comings I felt it had was what if you want to check if it contains a certain error. But perhaps that is not something to be done. Perhaps if you want one special error it would be much better to do something like:

```go
slice := errors.NewErrorSlice()
// add errors
return errors.WrapError(mySpecialErr, slice)
```

In which case perhaps the type is good enough.

That reminds me:

Implement a `WrapError` method, in case we want to wrap errors like this;

Also you expose the `ErrorSlice` type, which I guess is fine. But than you should make it work as a nil type without panics and errors. What is mean is that the following should (perhaps) work:

```
var errs ErrorSlice
// ...
errs.Add(myError)
// ...
return errs
```

---

`errors.WrapError(mySpecialErr, slice)` should return an error that has `mySpecialErr` as cause and slice.Error() as  annotation, or...?

---

No it would be a new struct `withError`, which also implement the `causer` and `Error` interface. Which would simply contain 2 errors (the cause and the other error). This way it doesn't have to do any work to convert the Error in a string, when it isn't needed. Because it's possible that the user only wants to use this error for its Cause in the end. So would be silly to have to do all that string conversion work for nothing.